File: api/app/services/quota.py
"""quota 모듈 — Gemini API quota 초과 감지 + 수익화 W3 플랜/사용량 조회.

W9 Day 4·6·7: is_quota_exhausted — Gemini API 초과 감지 유틸리티.
  - class name 화이트리스트 + status code attribute + 메시지 fallback 3단계.
  - stdlib only, google SDK 직접 import 없이 type name 검사로 우회.

W3 수익화: plans/subscriptions(마이그 022) 를 읽어 유효 플랜 한도 산출 (read-only).
  - enforcement 는 rate_limit.py 통합 게이트(Task 4) 담당.
  - 어떤 DB 실패든 None 반환 (fail-open — 호출측이 quota skip + warning).
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from datetime import datetime, timezone

from app.db import get_supabase_client

logger = logging.getLogger(__name__)

# 알려진 quota 초과 exception class name 화이트리스트.
# - ResourceExhausted: google.api_core.exceptions (gRPC 표준)
# - ClientError: google.genai.errors (HTTP wrapper) — code 검사 필요
_QUOTA_EXCEPTION_NAMES: frozenset[str] = frozenset({
    "ResourceExhausted",
    "TooManyRequests",
})


def is_quota_exhausted(error_or_msg) -> bool:
    """quota 초과 케이스 감지 — Exception object 또는 str message 모두 수용.

    감지 우선순위 (W9 Day 7 — 한계 #50 회수)
    1. exception class name 화이트리스트 — google SDK 의 표준 type
    2. exception 의 status_code/code attribute == 429
    3. 메시지 휴리스틱 (RESOURCE_EXHAUSTED / 429 / QUOTA) — fallback

    1·2 는 SDK 응답 형식과 무관하게 정확. 3 은 문자열 변경 시 회귀 가능하나,
    1·2 가 false negative 일 때 안전망 역할.
    """
    if isinstance(error_or_msg, BaseException):
        # 1) class name 직접 검사 — import 없이도 google SDK 표준 type 인식
        if type(error_or_msg).__name__ in _QUOTA_EXCEPTION_NAMES:
            return True
        # 2) HTTP-style status code attribute (google.genai.errors.ClientError 등)
        for attr in ("status_code", "code"):
            value = getattr(error_or_msg, attr, None)
            if value == 429:
                return True
        # 3) fallback — 메시지 검사
        msg = str(error_or_msg)
    else:
        msg = error_or_msg

    if not msg:
        return False
    upper = msg.upper()
    return (
        "RESOURCE_EXHAUSTED" in upper
        or "429" in msg
        or "QUOTA" in upper
    )


# ---------------------------------------------------------------------------
# W3 수익화 — 플랜(Free/Pro) 해석 + 사용량 카운트 조회
# ---------------------------------------------------------------------------

_EFFECTIVE_STATUSES = ("active", "past_due")


@dataclass(frozen=True)
class PlanLimits:
    code: str
    max_documents: int
    answers_per_day: int


@dataclass(frozen=True)
class SubscriptionView:
    plan_code: str
    status: str  # active | past_due | canceled | none(행 없음)
    current_period_end: str | None


def get_subscription_view(user_id: str) -> SubscriptionView:
    """구독 표시용 (/me/subscription). 행 없음/실패 → free·none (fail-open)."""
    try:
        rows = (
            get_supabase_client()
            .table("subscriptions")
            .select("plan_code, status, current_period_end")
            .eq("user_id", user_id)
            .limit(1)
            .execute()
            .data
        ) or []
        if not rows:
            return SubscriptionView(plan_code="free", status="none", current_period_end=None)
        r = rows[0]
        return SubscriptionView(
            plan_code=r.get("plan_code", "free"),
            status=r.get("status", "none"),
            current_period_end=r.get("current_period_end"),
        )
    except Exception as exc:  # noqa: BLE001
        logger.warning("구독 조회 실패 (user=%s): %s", user_id, exc)
        return SubscriptionView(plan_code="free", status="none", current_period_end=None)


def get_effective_plan(user_id: str) -> PlanLimits | None:
    """유저의 유효 플랜 한도. 실패 시 None (fail-open).

    정책
    - 구독 행 없음 / status='canceled' → free.
    - status IN ('active', 'past_due') → 해당 plan_code (past_due = grace, W5-6 예약).
    - 어떤 DB 실패든 None 반환 (fail-open — 호출측이 quota skip + warning).
    """
    try:
        client = get_supabase_client()
        sub_rows = (
            client.table("subscriptions")
            .select("plan_code, status")
            .eq("user_id", user_id)
            .limit(1)
            .execute()
            .data
        ) or []
        code = "free"
        if sub_rows and sub_rows[0].get("status") in _EFFECTIVE_STATUSES:
            code = sub_rows[0]["plan_code"]

        plan_rows = (
            client.table("plans")
            .select("code, max_documents, answers_per_day")
            .eq("code", code)
            .limit(1)
            .execute()
            .data
        ) or []
        if not plan_rows:
            logger.warning("plans 테이블에 code=%s 없음 — quota fail-open", code)
            return None
        row = plan_rows[0]
        return PlanLimits(
            code=row["code"],
            max_documents=int(row["max_documents"]),
            answers_per_day=int(row["answers_per_day"]),
        )
    except Exception as exc:  # noqa: BLE001 — 조회 실패는 fail-open
        logger.warning("플랜 조회 실패 — quota fail-open (user=%s): %s", user_id, exc)
        return None


def count_active_documents(user_id: str) -> int | None:
    """보유 문서 수 (deleted_at IS NULL). 실패 시 None (fail-open)."""
    try:
        # limit(1) = payload 최소화. count="exact" 는 limit 과 무관하게 전체 건수를 resp.count 로 반환.
        resp = (
            get_supabase_client()
            .table("documents")
            .select("id", count="exact")
            .eq("user_id", user_id)
            .is_("deleted_at", "null")
            .limit(1)
            .execute()
        )
        return int(resp.count or 0)
    except Exception as exc:  # noqa: BLE001
        logger.warning("문서 수 카운트 실패 — quota fail-open (user=%s): %s", user_id, exc)
        return None


def get_todays_count(user_key: str, metric: str) -> int:
    """usage_counters 의 금일 카운트 (표시용 — /me/plan). 실패 시 0.

    enforcement 에 사용 금지 — 실패 시 0 이라 장애가 '사용량 없음' 으로 보인다.
    """
    try:
        today = datetime.now(timezone.utc).date().isoformat()
        rows = (
            get_supabase_client()
            .table("usage_counters")
            .select("count")
            .eq("user_key", user_key)
            .eq("metric", metric)
            .eq("period_date", today)
            .limit(1)
            .execute()
            .data
        ) or []
        return int(rows[0]["count"]) if rows else 0
    except Exception as exc:  # noqa: BLE001
        logger.warning("금일 사용량 조회 실패 — 0 반환 (key=%s): %s", user_key, exc)
        return 0

File: supabase/functions/_shared/me/quota.ts
/**
 * 플랜·사용량 조회 — `services/quota.py` 중 `/me/*` 가 쓰는 부분 포팅.
 *
 * ## 전부 fail-open 이다
 * DB 가 흔들려도 사용자를 막지 않는다. 실패 시 `getEffectivePlan` 은 `null`,
 * `storageUsedBytes`·`visionPagesUsedMonth` 는 `null`, `getTodaysCount` 는 `0`,
 * `getSubscriptionView` 는 free·none 을 돌려준다. **이 값들로 제한을 걸면 안 된다** —
 * 장애가 "사용량 없음" 으로 보이기 때문이다(원본 주석의 경고를 그대로 옮긴다).
 *
 * ## 2026-09-15 계량 교체 (스펙 §4 S4)
 * 문서 수 상한(plans 컬럼과 문서 카운트 함수)을 버리고 **저장 용량 + 월 Vision 페이지**로
 * 옮겼다. 자동 수집은 문서 수로 재면 첫 동기화에서 막히고, 실제 비용은
 * 용량과 Vision 페이지(페이지당 $0.005~0.03 실측)에서 난다.
 *
 * ## 날짜 기준이 `/stats` 와 다르다
 * `getTodaysCount` 는 **UTC 날짜**를 쓴다. `/stats` 의 월·주 집계와 vision 사용량은
 * KST 다. 한 응답 안에 두 기준이 섞여 있는 게 원본 상태이고, 여기서 통일하면 값이 바뀐다.
 */

import type { SupabaseClient } from "@supabase/supabase-js";

/** `active` 와 `past_due` 만 유효 — `past_due` 는 유예 기간이라 플랜을 유지한다. */
const EFFECTIVE_STATUSES = new Set(["active", "past_due"]);

export interface PlanLimits {
  code: string;
  answers_per_day: number;
  storage_bytes_limit: number;
  vision_pages_per_month: number;
}

export interface SubscriptionView {
  plan_code: string;
  /** `active` · `past_due` · `canceled` · `none`(행 없음). */
  status: string;
  current_period_end: string | null;
}

const FREE_VIEW: SubscriptionView = {
  plan_code: "free",
  status: "none",
  current_period_end: null,
};

/** 구독 표시용. 행이 없거나 조회가 실패하면 free·none. */
export async function getSubscriptionView(
  client: SupabaseClient,
  userId: string,
): Promise<SubscriptionView> {
  try {
    const { data, error } = await client
      .from("subscriptions")
      .select("plan_code, status, current_period_end")
      .eq("user_id", userId)
      .limit(1);
    if (error) throw new Error(error.message);
    const rows = (data ?? []) as Record<string, unknown>[];
    if (rows.length === 0) return { ...FREE_VIEW };
    const r = rows[0];
    return {
      // 원본은 `r.get("plan_code", "free")` — 키가 **없을 때만** 기본값이다.
      // 값이 `null` 이면 `null` 이 그대로 나간다.
      plan_code: ("plan_code" in r ? r.plan_code : "free") as string,
      status: ("status" in r ? r.status : "none") as string,
      current_period_end: (r.current_period_end ?? null) as string | null,
    };
  } catch (e) {
    console.warn(`구독 조회 실패 (user=${userId}):`, e);
    return { ...FREE_VIEW };
  }
}

/**
 * 유효 플랜 한도. 조회 실패는 `null`(fail-open).
 *
 * 구독 행이 없거나 `canceled` 면 `free`, `active`/`past_due` 면 그 `plan_code` 를 쓴다.
 * 그 코드로 `plans` 를 한 번 더 조회하는데, **거기 행이 없으면 `null`** 이다 —
 * 플랜 테이블이 비어 있으면 조용히 free 로 떨어지지 않고 호출부가 503 을 낸다.
 */
export async function getEffectivePlan(
  client: SupabaseClient,
  userId: string,
): Promise<PlanLimits | null> {
  try {
    const { data: subData, error: subErr } = await client
      .from("subscriptions")
      .select("plan_code, status")
      .eq("user_id", userId)
      .limit(1);
    if (subErr) throw new Error(subErr.message);
    const subRows = (subData ?? []) as Record<string, unknown>[];

    let code = "free";
    if (subRows.length > 0 && EFFECTIVE_STATUSES.has(subRows[0].status as string)) {
      code = subRows[0].plan_code as string;
    }

    const { data: planData, error: planErr } = await client
      .from("plans")
      .select("code, answers_per_day, storage_bytes_limit, vision_pages_per_month")
      .eq("code", code)
      .limit(1);
    if (planErr) throw new Error(planErr.message);
    const planRows = (planData ?? []) as Record<string, unknown>[];
    if (planRows.length === 0) {
      console.warn(`plans 테이블에 code=${code} 없음 — quota fail-open`);
      return null;
    }
    const row = planRows[0];
    return {
      code: row.code as string,
      answers_per_day: Math.trunc(Number(row.answers_per_day)),
      storage_bytes_limit: Math.trunc(Number(row.storage_bytes_limit)),
      vision_pages_per_month: Math.trunc(Number(row.vision_pages_per_month)),
    };
  } catch (e) {
    console.warn(`플랜 조회 실패 — quota fail-open (user=${userId}):`, e);
    return null;
  }
}

/** 보유 용량(바이트). 실패는 `null` — 제한 판정은 fail-open. */
export async function storageUsedBytes(
  client: SupabaseClient,
  userId: string,
): Promise<number | null> {
  try {
    const { data, error } = await client.rpc("storage_bytes_used", { p_user_id: userId });
    if (error) throw new Error(error.message);
    return Math.trunc(Number(data ?? 0));
  } catch (e) {
    console.warn(`storage_bytes_used 실패 — fail-open (user=${userId}):`, e);
    return null;
  }
}

/** 이번 달(KST) 1일 00:00 의 ISO. `/stats` 의 월 집계와 같은 기준이다. */
export function kstMonthStartIso(nowMs: number): string {
  const kst = new Date(nowMs + 9 * 3600_000);
  const start = Date.UTC(kst.getUTCFullYear(), kst.getUTCMonth(), 1) - 9 * 3600_000;
  return new Date(start).toISOString();
}

/** 이번 달 Vision 페이지 수. 실패는 `null`. */
export async function visionPagesUsedMonth(
  client: SupabaseClient,
  userId: string,
  nowMs: number,
): Promise<number | null> {
  try {
    const { data, error } = await client.rpc("vision_pages_used_since", {
      p_user_id: userId,
      p_since: kstMonthStartIso(nowMs),
    });
    if (error) throw new Error(error.message);
    return Math.trunc(Number(data ?? 0));
  } catch (e) {
    console.warn(`vision_pages_used_since 실패 — fail-open (user=${userId}):`, e);
    return null;
  }
}

export interface QuotaSettings {
  authEnabled: boolean;
  quotaEnforcementEnabled: boolean;
  ownerUserId: string | null;
}

/**
 * 플랜 quota 가 이 사용자에게 걸리는가 — **판정의 단일 소스**다.
 *
 * `rate_limit.ts`(답변 일일 한도)·`makeStorageCheck`(용량)·`ingest/quota_gate.ts`(월 Vision
 * 페이지)가 전부 이 함수를 쓴다. 규칙이 갈라지면 같은 사용자가 경로마다 다른 대접을 받는다.
 */
export function quotaActiveFor(
  user: { userId: string; isAuthenticated: boolean },
  s: QuotaSettings,
): boolean {
  return s.authEnabled && s.quotaEnforcementEnabled && user.isAuthenticated &&
    user.userId !== (s.ownerUserId ?? "");
}

/** persist 에 꽂는 용량 검사. quota 비활성·조회 실패는 `null`(fail-open). */
export function makeStorageCheck(
  client: SupabaseClient,
  user: { userId: string; isAuthenticated: boolean },
  s: QuotaSettings,
): (sizeBytes: number) => Promise<{ allowed: boolean; usedBytes: number; limitBytes: number } | null> {
  return async (sizeBytes) => {
    if (!quotaActiveFor(user, s)) return null;
    const plan = await getEffectivePlan(client, user.userId);
    if (plan === null || plan.storage_bytes_limit <= 0) return null;
    const used = await storageUsedBytes(client, user.userId);
    if (used === null) return null;
    return {
      allowed: used + sizeBytes <= plan.storage_bytes_limit,
      usedBytes: used,
      limitBytes: plan.storage_bytes_limit,
    };
  };
}

/** `YYYY-MM-DD` (**UTC**). `/stats` 의 KST 기준과 다르다 — 원본 그대로다. */
export function utcTodayIso(nowMs: number): string {
  return new Date(nowMs).toISOString().slice(0, 10);
}

/** 금일 사용량(표시용). 실패는 `0` — **제한 판정에 쓰면 안 된다.** */
export async function getTodaysCount(
  client: SupabaseClient,
  userKey: string,
  metric: string,
  nowMs: number,
): Promise<number> {
  try {
    const { data, error } = await client
      .from("usage_counters")
      .select("count")
      .eq("user_key", userKey)
      .eq("metric", metric)
      .eq("period_date", utcTodayIso(nowMs))
      .limit(1);
    if (error) throw new Error(error.message);
    const rows = (data ?? []) as { count: number }[];
    return rows.length ? Math.trunc(Number(rows[0].count)) : 0;
  } catch (e) {
    console.warn(`금일 사용량 조회 실패 — 0 반환 (key=${userKey}):`, e);
    return 0;
  }
}
